### PR TITLE
move state into variables

### PR DESCRIPTION
**Pull request by rc**
Addresses #378, overall state handling interface clean-up.

### CODE REVIEW

**Comment by rc** [comment]
It still remains to remove `sfepy/discrete/state.py` - `State.get_weighted_norm()` seems useful, so I will probably move it into `sfepy.linalg` as a function.

**Comment by rc** [comment]
TODO: list the function name/API changes here, and post it to the mailing list.

**Comment by rc** [comment]
Finally, I decided to not preserve the `State.get_weighted_norm()` and just removed the file.

**Comment by rc** [comment]
Description of the changes:

Old API:

- State instance holds a global DOF vector that contains as contiguous
  sub-vectors (slices) the data of the unknown (state) variables from the
  Variables container.
- Possible to set DOF vector to a variable without changing the DOF vector in
  the State instance that contains the variable. -> Prone to hard-to-debug bugs
  (#378)
- Inconsistent naming of functions (historical code evolution).

New API:

- No State, only Variables.
- Once Variables.init_state() is called, the state variables are locked -
  cannot use variable.set_data() directly.
- Problem.solve() and other functions that worked/returned State instance
  return now the Variables instance instead.
- Declarative problem description API not affected.
- Imperative problem description API changed in a not backward-compatible way
  due to:
  - Renamed or removed functions (mostly Variables and related Equations
    methods and former State methods).
  - Locking of the state variables.
  - No need to create State instances (no State class).
  - Problem.create_state() is now an alias to Problem.get_initial_state(),
    which sets up initial conditions - those have to be set using
    Problem.set_ics() before Problem.create_state()/.get_initial_state() are
    called.
  - Stationary examples likely affected only in post-processing of results,
    after calling Problem.solve().
  - Time dependent examples need to be adjusted already when dealing with the
    initial state: calling Problem.get_initial_state() should do the job
    (provided the conditions are set, see above).

Old API - list of functions:
```
State:
 .__call__ (def)
 .__init__ (def)
 .apply_ebc (def)
 .apply_ic (def)
 .copy (def)
 .create_output_dict (def)
 .fill (def)
 .from_variables (def)
 .get_parts (def)
 .get_reduced (def)
 .get_vec (def)
 .get_weighted_norm (def)
 .has_ebc (def)
 .init_history (def)
 .set_full (def)
 .set_parts (def)
 .set_reduced (def)
 .set_vec (def)
```
```
Variables (state-related subset and external DOF vector manipulation helpers):
 .advance (def)
 .apply_ebc (def)
 .apply_ic (def)
 .check_vector_size (def)
 .create_state_vector (def)
 .create_stripped_state_vector (def)
 .get_indx (def)
 .get_state_part_view (def)
 .get_state_parts (def)
 .has_ebc (def)
 .init_history (def)
 .make_full_vec (def)
 .set_data (def)
 .set_from_state (def)
 .set_state_part (def)
 .state_to_output (def)
 .strip_state_vector (def)
```
New API - list of functions:
```
Variables (state-related subset and external DOF vector manipulation helpers):
 .__call__ (def)
 .advance (def)
 .apply_ebc (def)
 .apply_ic (def)
 .check_vec_size (def)
 .create_output (def)
 .create_reduced_vec (def)
 .create_vec (def)
 .fill_state (def)
 .get_indx (def)
 .get_reduced_state (def)
 .get_state (def)
 .get_state_parts (def)
 .get_vec_part (def)
 .has_ebc (def)
 .init_history (def)
 .init_state (def)
 .invalidate_evaluate_caches (def)
 .make_full_vec (def)
 .reduce_vec (def)
 .set_data (def)
 .set_full_state (def)
 .set_reduced_state (def)
 .set_state (def)
 .set_state_parts (def)
 .set_vec_part (def)
```

File: sfepy/homogenization/coefs_phononic.py [hunk 1]
```diff
@@ -666,11 +666,11 @@ def post_process(self, eigs, mtx_s_phi, mtx_dib, problem):
 
         mtx_s_phi_schur = - sc.dot(mtx_dib, mtx_s_phi)
         aux = nm.empty((variables.adi.ptr[-1],), dtype=nm.float64)
-        set = variables.set_state_part
+        setv = variables.set_vec_part
         for ii in range(n_eigs):
-            set(aux, mtx_s_phi[:,ii], primary_var, stripped=True)
-            set(aux, mtx_s_phi_schur[:,ii], eliminated_var,
-                stripped=True)
+            setv(aux, primary_var, mtx_s_phi[:,ii], stripped=True)
```
Comment (vlukes) [inline sfepy/homogenization/coefs_phononic.py:672]: `stripped=True` --> `reduced=True` ?

File: doc/faq.rst [hunk 2]
```diff
@@ -113,11 +113,13 @@ Miscellaneous
    tangent matrix ``K``, the imperative API can be used as follows::
 
      # pb is a Problem instance,
+     pb.set_bcs(ebcs=Conditions([...])) #
```
Comment (vlukes) [inline doc/faq.rst:116]: remove ` #`

File: sfepy/homogenization/coefs_phononic.py [hunk 3]
```diff
@@ -666,11 +667,11 @@ def post_process(self, eigs, mtx_s_phi, mtx_dib, problem):
...
+            setv(aux, eliminated_var, mtx_s_phi_schur[:,ii],
+                 stripped=True)
```
Comment (vlukes) [inline sfepy/homogenization/coefs_phononic.py:674]: `stripped=True` --> `reduced=True` ?

File: examples/navier_stokes/navier_stokes.py [hunk 4]
```diff
@@ -173,9 +173,9 @@ def verify_incompressibility( out, problem, state, extend = False ):
     incompressibility condition."""
     from sfepy.base.base import nm, output, assert_
 
-    vv = problem.get_variables()
+    vv = state
     one = nm.ones( (vv['p'].field.n_nod,), dtype = nm.float64 )
-    vv['p'].set_data(one)
+    vv.set_state_parts({'p' : one})
```
[the remark this answers is not in the document]
Comment (rc) [reply examples/navier_stokes/navier_stokes.py:178]: Just old code, simplified now. Both u, p are in state, so no need to pass them as arguments.

File: sfepy/homogenization/coefs_phononic.py (hunk 1 above)
Comment (rc) [reply sfepy/homogenization/coefs_phononic.py:672]: Yes

File: doc/faq.rst (hunk 2 above)
Comment (rc) [reply doc/faq.rst:116]: Fixed (added text instead).

File: sfepy/homogenization/coefs_phononic.py (hunk 3 above)
Comment (rc) [reply sfepy/homogenization/coefs_phononic.py:674]: OK

File: examples/navier_stokes/navier_stokes.py [hunk 5]
```diff
@@ -173,9 +173,9 @@ def verify_incompressibility( out, problem, state, extend = False ):
     incompressibility condition."""
     from sfepy.base.base import nm, output, assert_
 
-    vv = problem.get_variables()
+    vv = state
```
[the remark this answers is not in the document]
Comment (rc) [reply examples/navier_stokes/navier_stokes.py:176]: OK.

File: examples/diffusion/poisson_parallel_interactive.py [hunk 6]
```diff
@@ -340,6 +339,7 @@ def solve_problem(mesh_filename, options, comm):
     output('saving solution...')
     timer.start()
 
+    u_i.locked = False
```
[the remark this answers is not in the document]
Comment (rc) [reply examples/diffusion/poisson_parallel_interactive.py]: Ones `Variables.init_state()` is called, `.set_data()` cannot be used on state variables as they are locked. So here the lock is removed in postprocessing to allow using `.set_data()` But I rather used `variables.set_state(sol0_i)` instead.

**Comment by rc** [comment]
Thanks for the review! Do you actually use SchurEVP anywhere (I do not)?

**Comment by vlukes** [comment]
> Do you actually use SchurEVP anywhere (I do not)?

No, I do not.

**Comment by rc** [comment]
I have actually found SchurEVP use in a very old code (= not runnable) for computing acoustic band gaps in a piezo-elastic material. So let's keep it, I may update the example and include it.

Otherwise all clear, or some questions remain?

**Comment by vlukes** [comment]
> Otherwise all clear, or some questions remain?

Everything is clear! You can merge.

**Pull request merged**